include/etl/signal.h

```diff
@@ -0,0 +1,250 @@
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions :
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+******************************************************************************/
+
+#ifndef ETL_SIGNAL_INCLUDED
+#define ETL_SIGNAL_INCLUDED
+
+#include <cstddef>
+
+#include "exception.h"
+#include "error_handler.h"
+#include "delegate.h"
+#include "platform.h"
+
+#if defined(ETL_USING_STL)
+  #include <algorithm>
+  #include <iterator>
+
+#if defined(ETL_CPP11_NOT_SUPPORTED)
+    #include "algorithm.h"
+#endif // ETL_CPP11_NOT_SUPPORTED
+#else 
+  #include "algorithm.h"
+  #include "iterator.h"
+#endif // ETL_USING_STL
+
+//*****************************************************************************
+///\defgroup signal signal
+/// Container that handles storing and calling callbacks.
+///\ingroup containers
+//*****************************************************************************
+
+namespace etl
+{
+  //***************************************************************************
+  ///\ingroup signal
+  /// The base class for signal exceptions.
+  //***************************************************************************
+  class signal_exception : public exception
+  {
+  public:
+    signal_exception(string_type reason_, string_type file_name_, numeric_type line_number_) 
+      : exception{reason_, file_name_, line_number_}
+    {
+    }
+  };
+
+  class signal_full final : public signal_exception
+  {
+  public:
+    signal_full(string_type file_name_, numeric_type line_number_)
+      : signal_exception{ETL_ERROR_TEXT("signal:full", "A"), file_name_, line_number_}
+    {
+    }
+  };
+
+
+  //***************************************************************************
+  ///\ingroup signal
+  ///
+  ///\brief A lightweight signal class designed for efficient memory usage and 
+  /// ability to store in ROM.
+  ///
+  ///\tparam T: Callback signature.
+  ///\tparam LENGTH: Maximum number of slots that can be connected to the signal.
+  ///\tparam TSlot: Function-object type or container type that can be invoked.
+  ///
+  ///\todo Support for return types other than void (aggregate etc.)
+  //***************************************************************************
+  template <typename T, size_t LENGTH, typename TSlot = etl::delegate<T>>
+  class signal final
```

Comment:
If this class is meant to support C++03 then you should use ETL_FINAL.

You can check basic compatibility from C++03 to C++20 by running the script `etl/test/run-syntax-checks.sh` on Linux. You need both GCC and clang installed.

Reply:
My primary goal was to be c++14 compliant since that is what I use, but I tried to be as lower compliant as possible.  I didn't know that script existed, thanks, I'll run it and see what I find :)